connection.go

```diff
@@ -218,8 +206,11 @@ func (mc *mysqlConn) interpolateParams(query string, args []driver.Value) (strin
 	buf, err := mc.buf.takeCompleteBuffer()
 	if err != nil {
 		// can not take the buffer. Something must be wrong with the connection
+		mc.cleanup()
+		// interpolateParams would be called before sending any query.
+		// So its safe to retry.
 		mc.log(err)
-		return "", ErrInvalidConn
+		return "", driver.ErrBadConn
```

Comment:
This is simply a bug fix. Since it did not close the connection and also did not return ErrBadConn, if the connection was in a state where takeCompleteBuffer() failed for any reason, ErrInvalidConn could be returned repeatedly, depending on how the application was used.
This fix may need to be backported. I will create a separate PR.